### DISCUSSION TITLE
Doesn't Work Save To File

### DISCUSSION THREAD

**Discussion by vshlykov** [coqui-ai/TTS#2988] [General Q&A]
I try to use simple example from docs and can't to save speech to file. This is my code:
```
import torch
from TTS.api import TTS

tts_model = TTS(model_name="tts_models/multilingual/multi-dataset/xtts_v1", progress_bar=True)

device = 'cuda'
tts_model.to(device)

text = "This is a test! This is also a test!!"

output_file_path = "e:/whisper/output.wav"

tts_model.tts_to_file(text=text, language="en", file_path=output_file_path)
```

And this is errors:

PS E:\whisper> python tts.py
 > tts_models/multilingual/multi-dataset/xtts_v1 is already downloaded.
 > Using model: xtts
 > Text splitted to sentences.
['This is a test!', 'This is also a test!!']
Traceback (most recent call last):
  File "E:\whisper\tts.py", line 18, in <module>
    tts_model.tts_to_file(text=text, language="en", file_path=output_file_path)
  File "E:\whisper\TTS\TTS\api.py", line 390, in tts_to_file
    wav = self.tts(text=text, speaker=speaker, language=language, speaker_wav=speaker_wav, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\whisper\TTS\TTS\api.py", line 337, in tts
    wav = self.synthesizer.tts(
          ^^^^^^^^^^^^^^^^^^^^^
  File "E:\whisper\TTS\TTS\utils\synthesizer.py", line 375, in tts
    outputs = self.tts_model.synthesize(
              ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\whisper\TTS\TTS\tts\models\xtts.py", line 428, in synthesize
    return self.inference_with_config(text, config, ref_audio_path=speaker_wav, language=language, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\whisper\TTS\TTS\tts\models\xtts.py", line 450, in inference_with_config
    return self.inference(text, ref_audio_path, language, **settings)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\idres\AppData\Local\Programs\Python\Python311\Lib\site-packages\torch\utils\_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "E:\whisper\TTS\TTS\tts\models\xtts.py", line 538, in inference
    ) = self.get_conditioning_latents(audio_path=ref_audio_path, gpt_cond_len=gpt_cond_len)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\whisper\TTS\TTS\tts\models\xtts.py", line 401, in get_conditioning_latents
    gpt_cond_latents = self.get_gpt_cond_latents(audio_path, length=gpt_cond_len)  # [1, 1024, T]
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\whisper\TTS\TTS\tts\models\xtts.py", line 367, in get_gpt_cond_latents
    audio = load_audio(audio_path)
            ^^^^^^^^^^^^^^^^^^^^^^
  File "E:\whisper\TTS\TTS\tts\models\xtts.py", line 31, in load_audio
    audio, sampling_rate = torchaudio.load(audiopath)
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\idres\AppData\Local\Programs\Python\Python311\Lib\site-packages\torchaudio\backend\soundfile_backend.py", line 221, in load
    with soundfile.SoundFile(filepath, "r") as file_:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\idres\AppData\Local\Programs\Python\Python311\Lib\site-packages\soundfile.py", line 658, in __init__
    self._file = self._open(file, mode_int, closefd)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\idres\AppData\Local\Programs\Python\Python311\Lib\site-packages\soundfile.py", line 1212, in _open
    raise TypeError("Invalid file: {0!r}".format(self.name))
TypeError: Invalid file: None

Looks like problem with recording in file but I can't understand why.

**Comment by Xprogrammer777** [comment]
> I try to use simple example from docs and can't to save speech to file. This is my code:
> 
> ```
> import torch
> from TTS.api import TTS
> 
> tts_model = TTS(model_name="tts_models/multilingual/multi-dataset/xtts_v1", progress_bar=True)
> 
> device = 'cuda'
> tts_model.to(device)
> 
> text = "This is a test! This is also a test!!"
> 
> output_file_path = "e:/whisper/output.wav"
> 
> tts_model.tts_to_file(text=text, language="en", file_path=output_file_path)
> ```
> 
> And this is errors:
> 
> PS E:\whisper> python tts.py
> 
> > tts_models/multilingual/multi-dataset/xtts_v1 is already downloaded.
> > Using model: xtts
> > Text splitted to sentences.
> > ['This is a test!', 'This is also a test!!']
> > Traceback (most recent call last):
> > File "E:\whisper\tts.py", line 18, in 
> > tts_model.tts_to_file(text=text, language="en", file_path=output_file_path)
> > File "E:\whisper\TTS\TTS\api.py", line 390, in tts_to_file
> > wav = self.tts(text=text, speaker=speaker, language=language, speaker_wav=speaker_wav, **kwargs)
> > ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> > File "E:\whisper\TTS\TTS\api.py", line 337, in tts
> > wav = self.synthesizer.tts(
> > ^^^^^^^^^^^^^^^^^^^^^
> > File "E:\whisper\TTS\TTS\utils\synthesizer.py", line 375, in tts
> > outputs = self.tts_model.synthesize(
> > ^^^^^^^^^^^^^^^^^^^^^^^^^^
> > File "E:\whisper\TTS\TTS\tts\models\xtts.py", line 428, in synthesize
> > return self.inference_with_config(text, config, ref_audio_path=speaker_wav, language=language, **kwargs)
> > ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> > File "E:\whisper\TTS\TTS\tts\models\xtts.py", line 450, in inference_with_config
> > return self.inference(text, ref_audio_path, language, **settings)
> > ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> > File "C:\Users\idres\AppData\Local\Programs\Python\Python311\Lib\site-packages\torch\utils_contextlib.py", line 115, in decorate_context
> > return func(*args, **kwargs)
> > ^^^^^^^^^^^^^^^^^^^^^
> > File "E:\whisper\TTS\TTS\tts\models\xtts.py", line 538, in inference
> > ) = self.get_conditioning_latents(audio_path=ref_audio_path, gpt_cond_len=gpt_cond_len)
> > ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> > File "E:\whisper\TTS\TTS\tts\models\xtts.py", line 401, in get_conditioning_latents
> > gpt_cond_latents = self.get_gpt_cond_latents(audio_path, length=gpt_cond_len)  # [1, 1024, T]
> > ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> > File "E:\whisper\TTS\TTS\tts\models\xtts.py", line 367, in get_gpt_cond_latents
> > audio = load_audio(audio_path)
> > ^^^^^^^^^^^^^^^^^^^^^^
> > File "E:\whisper\TTS\TTS\tts\models\xtts.py", line 31, in load_audio
> > audio, sampling_rate = torchaudio.load(audiopath)
> > ^^^^^^^^^^^^^^^^^^^^^^^^^^
> > File "C:\Users\idres\AppData\Local\Programs\Python\Python311\Lib\site-packages\torchaudio\backend\soundfile_backend.py", line 221, in load
> > with soundfile.SoundFile(filepath, "r") as file_:
> > ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> > File "C:\Users\idres\AppData\Local\Programs\Python\Python311\Lib\site-packages\soundfile.py", line 658, in **init**
> > self._file = self._open(file, mode_int, closefd)
> > ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> > File "C:\Users\idres\AppData\Local\Programs\Python\Python311\Lib\site-packages\soundfile.py", line 1212, in _open
> > raise TypeError("Invalid file: {0!r}".format(self.name))
> > TypeError: Invalid file: None
> 
> Looks like problem with recording in file but I can't understand why.

Had the same issue, I got a fix.
Try this : tts.tts_to_file("Hello world", file_path=r"output.wav", language="en", speaker_wav=[r"C:\Users\me\path\to\voice\voices\voice.wav"])
